BuildScript/BuildTasks/TestConfigurationsFactory.cs

```diff
@@ -0,0 +1,220 @@
+﻿// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+//
+// See the NOTICE file distributed with this work for additional information
+// regarding copyright ownership.  rubicon licenses this file to you under 
+// the Apache License, Version 2.0 (the "License"); you may not use this 
+// file except in compliance with the License.  You may obtain a copy of the 
+// License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software 
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT 
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the 
+// License for the specific language governing permissions and limitations
+// under the License.
+// 
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace Remotion.BuildScript.BuildTasks
+{
+  public class TestConfigurationsFactory
+  {
+    private readonly MetadataValueCollection _supportedBrowsers;
+    private readonly MetadataValueCollection _supportedDatabaseSystems;
+    private readonly MetadataValueCollection _supportedPlatforms;
+    private readonly MetadataValueCollection _supportedConfigurationIDs;
+    private readonly MetadataValueDictionary _supportedExecutionRuntimes;
+    private readonly MetadataValueCollection _supportedTargetRuntimes;
+
+    public TestConfigurationsFactory (
+        MetadataValueCollection supportedBrowsers,
+        MetadataValueCollection supportedDatabaseSystems,
+        MetadataValueCollection supportedPlatforms,
+        MetadataValueCollection supportedConfigurationIDs,
+        MetadataValueDictionary supportedExecutionRuntimes,
+        MetadataValueCollection supportedTargetRuntimes)
+    {
+      _supportedBrowsers = supportedBrowsers;
+      _supportedDatabaseSystems = supportedDatabaseSystems;
+      _supportedPlatforms = supportedPlatforms;
+      _supportedConfigurationIDs = supportedConfigurationIDs;
+      _supportedExecutionRuntimes = supportedExecutionRuntimes;
+      _supportedTargetRuntimes = supportedTargetRuntimes;
+    }
+
+    public IReadOnlyCollection<TestConfiguration> CreateTestConfigurations (
+        string testAssemblyFullPath,
+        IReadOnlyCollection<string> rawTestConfigurations,
+        string testSetupBuildFile = null)
+    {
+      var output = new List<TestConfiguration>();
+
+      foreach (var rawTestConfiguration in rawTestConfigurations)
+      {
+        var configurationItems = GetConfigurationItems (rawTestConfiguration);
```

Comment:
extract body and use linq Select / ToList ? Possibly, move the duplicate check into GetConfigurationItems and just chain GetCofnigurationItems and CreateTestConfiguration via two Select calls?